Implemented the Fruit Shop application following SOLID principle.

PavloMelnyk10 opened the pull request:
-Added model classes and storage
-Implemented file reading and writing
-Created data converter
-Implemented operation handlers and operation strategy -Added shop service for processing transactions
-Implemented report generation and writing to output file -Organized classes into packages

PavloMelnyk10 committed:
Implemented the Fruit Shop application following SOLID principle.
-Added model classes and storage
-Implemented file reading and writing
-Created data converter
-Implemented operation handlers and operation strategy
-Added shop service for processing transactions
-Implemented report generation and writing to output file
-Organized classes into packages

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/HelloWorld.java:
```diff
@@ -1,9 +1,52 @@
 public class HelloWorld {
-    // HINT: In the `public static void main(String[] args)` it is better to create instances of your classes, 
-    // and call their methods, but do not write any business logic in the `main` method!
+    public static void main(String[] args) {
```
Don't keep all logic in a single package. Your project structure should consist of different packages such as `db`, `model`, `service`, `service.impl`, `strategy`, etc. It seems that the 'HelloWorld' class is not correctly named according to its responsibilities and it's in the default package. Consider placing it in a suitably named package and renaming it to reflect its role in the application, such as 'MainApplication' or similar.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/report/ReportGeneratorImpl.java:
```diff
@@ -0,0 +1,18 @@
+import core.basesyntax.storage.Storage;
+import java.util.Map;
+
+public class ReportGeneratorImpl implements ReportGenerator {
```
According to the checklist, don't begin class or method implementation with an empty line. Remove the empty line at the start of the class implementation.

mateacademy-ai-mentor replied:
According to the checklist, classes should not begin with an empty line. Please remove the empty line before the class declaration.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/report/ReportGeneratorImpl.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public String getReport() {
+        StringBuilder report = new StringBuilder("fruit,quantity\n");
+        for (Map.Entry<String, Integer> entry : Storage.getFruitStorage().entrySet()) {
```
Follow the encapsulation principle. Consider making `Storage.getFruitStorage()` return a copy of the storage map or an unmodifiable view to prevent exposing the internal structure.

mateacademy-ai-mentor replied:
It's a good practice to follow the encapsulation principle. The `Storage.getFruitStorage()` method exposes the internal storage details. Instead, consider providing a method in `Storage` that returns a copy of the storage data or an unmodifiable view.

mateacademy-ai-mentor reviewed (changes requested):
It looks like you've done a great job on the implementation overall! 🌟 However, there are a few key areas that need adjustments to ensure the code is maintainable and adheres to best practices. Specifically, consider the encapsulation for the `Storage` class and the placement of the `Operation` enum. Also, make sure to validate inputs like the `transaction` object to prevent potential issues down the line. Keep up the good work, and with these changes, your code will be even better! 👍

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

PavloMelnyk10 committed:
Refactored project structure:
-Moved business logic from main to FruitShopController and FruitShopService
-Implemented dependency injection using Injector class
-Improved adherence to SOLID principles

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FruitShopApplication.java:
```diff
@@ -0,0 +1,12 @@
+import core.basesyntax.service.FruitShopController;
+
+public class FruitShopApplication {
+    public static void main(String[] args) {
+        FruitShopController controller = new FruitShopController();
+        controller.run();
+    }
```
Make all services independent and call them in the right order in `main()` method step by step. The current implementation directly calls a controller which abstracts the flow of the program. Instead, instantiate and use the independent services here, like FileReader, DataConverter, OperationStrategy, etc., and orchestrate them to perform the program's operations.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/converter/DataConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+public class DataConverterImpl implements DataConverter {
+    public List<FruitTransaction> convertToTransaction(List<String> data) {
```
The method `convertToTransaction` should not begin with an empty line according to the checklist.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/converter/DataConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+        return transactions;
+    }
+
+    private Operation mapToOperation(String code) {
```
The method `mapToOperation` should not begin with an empty line according to the checklist.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/converter/DataConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private Operation mapToOperation(String code) {
+        return switch (code) {
+            case "b" -> Operation.BALANCE;
+            case "s" -> Operation.SUPPLY;
+            case "p" -> Operation.PURCHASE;
+            case "r" -> Operation.RETURN;
+            default -> throw new IllegalArgumentException("Unknown operation code: " + code);
```
Avoid using switch-cases. It is recommended to use the Strategy pattern instead. This will make adding new operations easier and your code cleaner.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/converter/DataConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+    public List<FruitTransaction> convertToTransaction(List<String> data) {
+        List<FruitTransaction> transactions = new ArrayList<>();
+        for (String line : data) {
+            if (line.startsWith("type")) {
```
You have hardcoded the header check. Instead, consider using a more flexible solution that does not depend on the content of the file, like skipping the first line or using a flag to indicate a header is present.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/converter/DataConverterImpl.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private Operation mapToOperation(String code) {
+        return switch (code) {
```
Using the `switch` statement here contradicts the checklist's advice to avoid `switch-cases` and `if-else` constructions. Consider using a `Map<String, Operation>` to map operation codes to their corresponding `Operation` enum values.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/file/writer/FileWriterImpl.java:
```diff
@@ -0,0 +1,16 @@
+
+public class FileWriterImpl implements FileWriter {
+    @Override
+    public void write(String data, String filePath) {
```
Avoid hardcode in your solution. It's better to pass the file path as a parameter to the method, which you have done correctly. Ensure that the path provided is relative and not absolute, to maintain the portability of the code.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/file/writer/FileWriterImpl.java:
```diff
@@ -0,0 +1,16 @@
+        } catch (IOException e) {
+            throw new RuntimeException("Cannot write to file: " + filePath, e);
+        }
+    }
```
Follow the encapsulation principle. Consider making your class and its methods as restrictive as necessary. For example, if the class is not intended to be subclassed, declare it as final.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/BalanceOperationHandler.java:
```diff
@@ -0,0 +1,12 @@
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.storage.Storage;
+
+public class BalanceOperationHandler implements OperationHandler {
```
According to the checklist, there shouldn't be an empty line at the beginning of a class or method implementation. Please remove the redundant empty line after the package statement.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/PurchaseOperationHandler.java:
```diff
@@ -0,0 +1,16 @@
+    @Override
+    public void apply(FruitTransaction transaction) {
+        int currentQuantity = Storage.getFruitQuantity(transaction.getFruit());
+        if (currentQuantity < transaction.getQuantity()) {
+            throw new RuntimeException("Not enough stock for " + transaction.getFruit());
```
When handling the Purchase operation, ensure that after the transaction, the resulting balance of the fruit in the Storage must not be negative. If it is, throw a RuntimeException with an informative message. The current implementation correctly throws an exception if the stock is insufficient, but the message could be more informative by including the attempted quantity and the current stock available.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/PurchaseOperationHandler.java:
```diff
@@ -0,0 +1,16 @@
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.storage.Storage;
+
+public class PurchaseOperationHandler implements OperationHandler {
```
Don't begin class or method implementation with an empty line. The class `PurchaseOperationHandler` starts with an empty line which should be removed according to the checklist.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/ReturnOperationHandler.java:
```diff
@@ -0,0 +1,12 @@
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.storage.Storage;
+
+public class ReturnOperationHandler implements OperationHandler {
```
The class name `ReturnOperationHandler` is not following Java class naming conventions. Consider renaming it to match the pattern of `VerbNoun` format, which could be something like `HandleReturnOperation`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/ReturnOperationHandler.java:
```diff
@@ -0,0 +1,12 @@
+public class ReturnOperationHandler implements OperationHandler {
+    @Override
+    public void apply(FruitTransaction transaction) {
+        Storage.addFruit(transaction.getFruit(), transaction.getQuantity());
```
Directly accessing the `Storage` class and its methods from the `ReturnOperationHandler` violates the encapsulation principle. Consider adding a service layer that interacts with the `Storage` and use that service here instead.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/handler/impl/SupplyOperationHandler.java:
```diff
@@ -0,0 +1,12 @@
+public class SupplyOperationHandler implements OperationHandler {
+    @Override
+    public void apply(FruitTransaction transaction) {
+        Storage.addFruit(transaction.getFruit(), transaction.getQuantity());
```
The 'addFruit' method should be named according to its function. Since it's modifying the quantity of the fruit in the storage, a name like 'updateFruitQuantity' would be more descriptive and align with the checklist's recommendation for clear method naming.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/FruitTransaction.java:
```diff
@@ -0,0 +1,37 @@
+package core.basesyntax.model;
+
+public class FruitTransaction {
```
Don't begin class implementation with an empty line. Please remove the redundant empty line at the beginning of the class definition.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/model/FruitTransaction.java:
```diff
@@ -0,0 +1,37 @@
+    public void setQuantity(int quantity) {
+        this.quantity = quantity;
+    }
+}
```
Don't end class implementation with an empty line. Please remove the redundant empty line at the end of the class definition.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/FruitShopController.java:
```diff
@@ -0,0 +1,18 @@
+import core.basesyntax.util.Injector;
+
+public class FruitShopController {
+    private static final String INPUT_FILE_PATH = "src/main/resources/reportToRead.csv";
+    private static final String OUTPUT_FILE_PATH = "src/main/resources/finalReport.csv";
```
Avoid hardcode in your solution. It's better to pass file paths as parameters to the `run` method or to the constructor of `FruitShopController`, instead of using hardcoded paths. This will make your application more flexible and easier to test.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/FruitShopController.java:
```diff
@@ -0,0 +1,18 @@
+        this.fruitShopService = Injector.getFruitShopService();
+    }
+
+    public void run() {
+        fruitShopService.processTransactions(INPUT_FILE_PATH, OUTPUT_FILE_PATH);
```
Make sure to make all services independent and call them in the `main()` method. The `run` method here seems to be tightly coupling the `FruitShopService` with file paths. Consider refactoring this to follow the SOLID principles more closely.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/FruitShopServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.List;
+
+public class FruitShopServiceImpl implements FruitShopService {
+    private final FileReader fileReader;
+    private final FileWriter fileWriter;
+    private final DataConverter dataConverter;
+    private final OperationStrategy operationStrategy;
+    private final ReportGenerator reportGenerator;
```
According to the checklist, classes should not begin with an empty line. Please remove the empty line after the class declaration to follow the code style guidelines.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/FruitShopServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+        List<String> inputData = fileReader.read(inputFilePath);
+        List<FruitTransaction> transactions = dataConverter.convertToTransaction(inputData);
+
+        for (FruitTransaction transaction : transactions) {
+            OperationHandler handler = operationStrategy.getHandler(transaction.getOperation());
+            handler.apply(transaction);
```
Consider making your services more independent as per the checklist. Currently, the `processTransactions` method is doing multiple things: reading input data, converting it, processing transactions, and generating a report. It would be better to call these services in the `main()` method, making each service responsible for a single piece of functionality.

mateacademy-ai-mentor replied:
The checklist recommends using the Strategy pattern without switch-cases or if-else constructions. Ensure that the `operationStrategy` uses a map of strategies to avoid such constructions and adheres to the Strategy pattern correctly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/ShopServiceImpl.java:
```diff
@@ -0,0 +1,22 @@
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private final OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            OperationHandler handler = operationStrategy.getHandler(transaction.getOperation());
+            handler.apply(transaction);
+        }
+    }
+}
```
Ensure that the ShopServiceImpl class is placed in the correct package according to the task checklist. It should be in the 'service.impl' package since it is an implementation of a service.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/service/ShopServiceImpl.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private final OperationStrategy operationStrategy;
```
According to the encapsulation principle, consider making the 'operationStrategy' field private if it's not intended to be accessed directly from outside the class.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/storage/Storage.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Storage {
+    private static final Map<String, Integer> fruitStorage = new HashMap<>();
+
+    public static Map<String, Integer> getFruitStorage() {
+        return Collections.unmodifiableMap(fruitStorage);
+    }
+
+    public static void addFruit(String fruit, int quantity) {
+        fruitStorage.merge(fruit, quantity, Integer::sum);
+    }
+
+    public static void setFruitQuantity(String fruit, int quantity) {
+        fruitStorage.put(fruit, quantity);
+    }
+
+    public static int getFruitQuantity(String fruit) {
+        return fruitStorage.getOrDefault(fruit, 0);
+    }
+}
```
Consider adhering to the Single Responsibility Principle. While this `Storage` class is responsible for providing access to the fruit storage, the actual operations on the storage (addition, setting, getting) might be better encapsulated in a separate class that deals with storage manipulation. This would make the `Storage` class focused only on providing the storage without knowing how it's manipulated.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/util/Injector.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Injector {
+    public static FruitShopService getFruitShopService() {
+        Map<Operation, OperationHandler> operationHandlers = new HashMap<>();
+        operationHandlers.put(Operation.BALANCE, new BalanceOperationHandler());
+        operationHandlers.put(Operation.SUPPLY, new SupplyOperationHandler());
+        operationHandlers.put(Operation.PURCHASE, new PurchaseOperationHandler());
+        operationHandlers.put(Operation.RETURN, new ReturnOperationHandler());
+        OperationStrategy operationStrategy = new OperationStrategyImpl(operationHandlers);
+
+        DataConverter dataConverter = new DataConverterImpl();
+        ReportGenerator reportGenerator = new ReportGeneratorImpl();
+        FileReader fileReader = new FileReaderImpl();
+        FileWriter fileWriter = new FileWriterImpl();
+
+        return new FruitShopServiceImpl(fileReader, fileWriter, dataConverter,
+                operationStrategy, reportGenerator);
+    }
```
The Injector class is currently creating instances of services directly within the `getFruitShopService` method. This approach tightly couples the service implementations to this Injector class, which can make it harder to manage dependencies, especially if the application grows in complexity. Consider using a dependency injection framework like Spring or Guice, or at least abstracting the creation of these instances into separate factory methods to improve encapsulation and make the codebase more maintainable.

mateacademy-ai-mentor reviewed (approved):
Great job on following the guidelines and implementing the necessary classes and interfaces! Just remember to keep an eye on encapsulation and the Single Responsibility Principle as your application grows. Keep up the good work! 🌟

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨
